### PR TITLE
lxd: Add VMs support

### PR DESCRIPTION
**Pull request by CajuM**
###### Motivation for this change
LXD has support for virtual machines, but this support isn't present in nixpkgs.

###### Things done
- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by CajuM** [comment]
I did not update the tests because I can't upload a VM image to tarballs.nixos.org

File: pkgs/tools/admin/lxd/default.nix [hunk 1]
```diff
@@ -14,6 +15,38 @@ let
     [ iptables-nftables-compat ]
   else
     [ iptables ebtables ];
+  lxd-agent = buildGoPackage rec {
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/default.nix:18]:
```suggestion
    [ iptables ebtables ];

  lxd-agent = buildGoPackage rec {
```

File: pkgs/tools/admin/lxd/default.nix [hunk 2]
```diff
@@ -14,6 +15,38 @@ let
...
+    pname = "lxd-agent";
+    version = "4.8";
+
+    goPackagePath = "github.com/lxc/lxd";
+
+    buildFlags = [ "-ldflags=-extldflags=-static" "-tags libsqlite3" ];
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/default.nix:24]: Does it hurt to add ``-s -w``?

File: pkgs/tools/admin/lxd/default.nix [hunk 3]
```diff
@@ -14,6 +15,38 @@ let
...
+
+    src = fetchurl {
+      url = "https://github.com/lxc/lxd/releases/download/lxd-${version}/lxd-${version}.tar.gz";
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/default.nix:27]: Just a question: Whats the difference with those compared to the github tarball?

**Review by SuperSandro2000 — COMMENTED** [review]
is there a reason why you did not put this into a separate package?

**Comment by CajuM** [comment]
I did not think it necessary to have an independent lxd-agent available for installation. lxd takes care of mounting it on the guest. It also may not be wise to have a different version of lxd than lxd-agent.

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 105651` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>lxd</li>
  </ul>
</details>

**Comment by CajuM** [comment]
> Result of `nixpkgs-review pr 105651` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 1 package built:

Should I add lxd-agent to buildInputs? It seems to get built anyway and placed in the lxd package.
Besides this, is there anything else pending?

**Comment by CajuM** [comment]
Hello, sorry for the long delay.
In #114194 gptfdisk and SeaBIOS are added as dependencies. But, they don't seem to be used in LXD. LXD 4.13 does not support any other BIOS than OVMF, as far as I can tell.
I solved the merge conflict. Should I also add a useQemu flag?

File: pkgs/tools/admin/lxd/default.nix [hunk 4]
```diff
@@ -15,18 +16,49 @@ let
...
   };
 
+  lxd-agent = buildGoPackage rec {
+    inherit version src;
+
+    pname = "lxd-agent";
+
+    goPackagePath = "github.com/lxc/lxd";
+
+    buildFlags = [ "-ldflags=-extldflags=-static" "-ldflags=-s" "-ldflags=-w" "-tags libsqlite3" ];
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/default.nix:34]:
```suggestion
    buildFlags = [ "-ldflags=-extldflags=-static" "-ldflags=-s -w" "-tags libsqlite3" ];
```

**Comment by SuperSandro2000** [comment]
> Should I also add a useQemu flag?

I think that is a good idea because qemu is quite big.

**Comment by CajuM** [comment]
> > Should I also add a useQemu flag?
> 
> I think that is a good idea because qemu is quite big.

In that case we'll also need an extra config flag in the virtualisation.lxd module

File: pkgs/tools/admin/lxd/default.nix (hunk 4 above)
Comment (CajuM) [reply pkgs/tools/admin/lxd/default.nix:34]: This gives me a build error

**Comment by SuperSandro2000** [comment]
> In that case we'll also need an extra config flag in the virtualisation.lxd module.

Can we change the package there already?

**Comment by CajuM** [comment]
> > In that case we'll also need an extra config flag in the virtualisation.lxd module.
> 
> Can we change the package there already?

Yes, we can. My bad.

**Comment by astralbijection** [comment]
What's the status on this? I know it's marked as stale, but I'd be happy to see this feature merged. If needed, I'd be interested in contributing as well.

**Comment by astralbijection** [comment]
I'll try building off of this and submit a different PR.

**Comment by kellianoy** [comment]
Hello @CajuM ! Thank you for your work.
Is there a possibility that you also add virts-proxy-helper in the same manner you added the lxd-agent ?

**Pull request closed without merging**